File: server/dotnet/TinyBI.Engine/JsonModels/QueryJson.cs
using System.Collections.Generic;

namespace TinyBI.Engine.JsonModels
{
    public class QueryJson
    {
        public IList<string> Select { get; set; }

        public IList<AggregationJson> Aggregations { get; set; }

        public IList<FilterJson> Filters { get; set; }

        public IList<OrderingJson> OrderBy { get; set; }

        public bool Totals { get; set; }
    }

    public class QueryRecordJson
    {
        public IList<object> Selected { get; set; }

        public IList<object> Aggregated { get; set; }
    }

    public class QueryResultJson
    {
        public IList<QueryRecordJson> Records { get; set; }

        public QueryRecordJson Totals { get; set; }
    }
}

File: client/packages/tinybi/src/QueryJson.ts
// This is the wire protocol, matches the TinyBI.Engine declarations

export type AggregationType = "none" | "count" | "sum";

export interface AggregationJson {
    function: AggregationType;
    column: string;
    filters?: FilterJson[];
}

export interface OrderingJson {
    column: string;
    descending?: boolean;
}

export interface QueryJson {
    select?: string[];
    aggregations: AggregationJson[];
    filters?: FilterJson[];
    orderBy?: OrderingJson[];
    totals?: boolean;
}

export type FilterOperator = "=" | "<>" | ">" | "<" | ">=" | "<=";

export type FilterValue = string | number | boolean | Date | unknown;

export interface FilterJson {
    column: string;
    operator: FilterOperator;
    value: FilterValue;
}
